### PR TITLE
[REEF-1043] Add runtime name to evaluator descriptor c#

### PR DESCRIPTION
**Pull request by shulmanb**
This issue addressed by
     * Adding runtime name to the interface
     * Parsing runtime name from the serialized descriptor received from teh bridge

JIRA:
  [REEF-983](https://issues.apache.org/jira/browse/REEF-1043)

### CODE REVIEW

**Comment by afchung** [comment]
I'll take a look.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs [hunk 1]
```diff
@@ -68,6 +69,11 @@ public EvaluatorDescriptorImpl(string str)
                 }
                 settings.Add(pair[0], pair[1]);
             }
+            string runtimeName;
+            if (!settings.TryGetValue("RuntimeName", out runtimeName))
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs]: check runtimeName ``isNullOrWhitespace``.

File: lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs [hunk 2]
```diff
@@ -173,9 +173,30 @@ protected void ValidateSuccessForLocalRuntime(int numberOfEvaluatorsToClose, int
 
         protected void ValidateMessageSuccessfullyLogged(string message, string testFolder)
         {
-            string[] lines = File.ReadAllLines(GetLogFile(_stdout, testFolder));
-            string[] successIndicators = lines.Where(s => s.Contains(message)).ToArray();
-            Assert.IsTrue(successIndicators.Any());
+            string[] lines = null;
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs]: What is the purpose of this change?

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+using Org.Apache.REEF.Utilities;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Tests.Functional.Driver
+{
+    public class EvaluatorRequestingDriver : 
+        IObserver<IDriverStarted>, 
+        IObserver<IAllocatedEvaluator>,         
+        IObserver<IRunningTask>
+
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs]: nit: extra space.

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Interface;
+using Org.Apache.REEF.Tang.Util;
+using Org.Apache.REEF.Tests.Functional.Messaging;
+using Org.Apache.REEF.Utilities;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Tests.Functional.Driver
+{
+    public class EvaluatorRequestingDriver :
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs]: sealed.

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            LOGGER.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "submitting evaluator request"));
+            _evaluatorRequestor.Submit(request);
+            LOGGER.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "evaluator request submitted"));
+        }
+
+        public void OnNext(IAllocatedEvaluator eval)
+        {
+            LOGGER.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "Received evaluator from runtime: {0}.", eval.GetEvaluatorDescriptor().RuntimeName));
+            eval.Dispose();
+            /*string taskId = "Task_" + eval.Id;
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs]: Remove commented out code.

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                    .Build();
+            LOGGER.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "submitting evaluator request"));
+            _evaluatorRequestor.Submit(request);
+            LOGGER.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "evaluator request submitted"));
+        }
+
+        public void OnNext(IAllocatedEvaluator eval)
+        {
+            LOGGER.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "Received evaluator from runtime: {0}.", eval.GetEvaluatorDescriptor().RuntimeName));
+            eval.Dispose();
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs]: You've mentioned that this doesn't close Evaluator. Can you confirm that this is the case? I've checked the InterOp code and it does seem to call the Java ``close`` function correctly.

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            throw new NotImplementedException();
+        }
+
+        public void OnCompleted()
+        {
+            throw new NotImplementedException();
+        }
+
+        void IObserver<IAllocatedEvaluator>.OnError(Exception error)
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs]: Remove the explicit interface implementations.

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/RuntimeNameTask.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        public const string MessageSend = "MESSAGE:TASK";
+
+        private static readonly Logger LOGGER = Logger.GetLogger(typeof(RuntimeNameTask));
+
+        [Inject]
+        public RuntimeNameTask()
+        {
+        }
+
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/RuntimeNameTask.cs]: nit: extra space.

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/RuntimeNameTask.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+using System;
+using System.Threading;
+using Org.Apache.REEF.Common.Tasks;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Tests.Functional.Messaging
+{
+    public class RuntimeNameTask : ITask
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/RuntimeNameTask.cs]: sealed

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/RuntimeNameTask.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        public const string MessageSend = "MESSAGE:TASK";
+
+        private static readonly Logger LOGGER = Logger.GetLogger(typeof(RuntimeNameTask));
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/RuntimeNameTask.cs]: I know there is inconsistency in the C# codebase, but I believe first letter capitalization is the way to move forward with ``private static`` variables.

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/RuntimeNameTask.cs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+        [Inject]
+        public RuntimeNameTask()
+        {
+        }
+
+
+        public byte[] Call(byte[] memento)
+        {
+            Console.WriteLine("Hello, CLR TaskMsg!");
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/RuntimeNameTask.cs]: ``Logger.Log``?

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+namespace Org.Apache.REEF.Tests.Functional.Driver
+{
+    public class EvaluatorRequestingDriver : 
+        IObserver<IDriverStarted>, 
+        IObserver<IAllocatedEvaluator>,         
+        IObserver<IRunningTask>
+
+    {
+        private static readonly Logger LOGGER = Logger.GetLogger(typeof(EvaluatorRequestingDriver));
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs]: See other comment about naming.

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/RuntimeNameTest.cs [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+            CleanUp(testFolder);
+            TestRun(DriverConfigurationsWithEvaluatorRequest(), typeof(EvaluatorRequestingDriver), 1, "EvaluatorRequestingDriver", "local", testFolder);
+            ValidateMessageSuccessfullyLogged("Received evaluator from runtime: Local", testFolder);
+            CleanUp(testFolder);
+        }
+
+        public IConfiguration DriverConfigurationsWithEvaluatorRequest()
+        {
+            IConfiguration driverConfig = DriverConfiguration.ConfigurationModule
+            .Set(DriverConfiguration.OnDriverStarted, GenericType<EvaluatorRequestingDriver>.Class)
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/RuntimeNameTest.cs]: nit: indent the ``Set`` and ``Build``.

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs (hunk 5 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs]: Or, it seems like this is actually the code you wanted to submit.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs [hunk 14]
```diff
@@ -135,6 +142,10 @@ public string Rack
             get { return _rack; }
         }
 
+        public string RuntimeName
+        {
+            get { return _runtimeName; }
+        }
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs]: nit: space

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs [hunk 15]
```diff
@@ -39,15 +39,16 @@ public sealed class EvaluatorDescriptorImpl : IEvaluatorDescriptor
...
         {
             _nodeDescriptor = nodeDescriptor;
             _evaluatorType = type;
             _megaBytes = megaBytes;
             _core = core;
             _rack = rack;
+            _runtimeName = runtimeName;
         }
 
         /// <summary>
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs]: side note: not entirely relevant to this PR, but I was wondering if a ``static`` "factory" method is the better way to go here... I don't particularly like initializing the variables in two places, and we can't use ``this()`` since ``this()`` will call the ``internal`` constructor first.

**Comment by afchung** [comment]
@shulmanb Done with first pass. No major issues. Thanks!

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs (hunk 6 above)
Comment (shulmanb) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs]: I was wrong. It does close it..

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs (hunk 14 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs]: sorry I meant new line.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs (hunk 15 above)
Comment (shulmanb) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs]: We could do that. We will need to remove readonly from the members, and I don't really like it, but if someone will do that I won't strongly object

Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs]: We won't need to remove readonly from the members. The static function will just call ``new`` internally after parsing the ``string`` input.

File: lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs (hunk 2 above)
Comment (shulmanb) [inline lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs]: There is a race condition between file creation and reading. Changed it to be similar to the ValidateSuccessForLocalRuntime

**Comment by shulmanb** [comment]
Uploaded new iteration. Please take a look.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs (hunk 15 above)
Comment (shulmanb) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs]: I see, that would be fine though don't see any big problem with what we currently have.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs [hunk 16]
```diff
@@ -39,15 +39,16 @@ public sealed class EvaluatorDescriptorImpl : IEvaluatorDescriptor
...
-        internal EvaluatorDescriptorImpl(INodeDescriptor nodeDescriptor, EvaluatorType type, int megaBytes, int core,
-            string rack = DefaultRackName)
+        internal EvaluatorDescriptorImpl(INodeDescriptor nodeDescriptor, EvaluatorType type, int megaBytes, int core, string rack = DefaultRackName, string runtimeName = "")
         {
             _nodeDescriptor = nodeDescriptor;
             _evaluatorType = type;
             _megaBytes = megaBytes;
             _core = core;
             _rack = rack;
+            _runtimeName = runtimeName;
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs]: can this actually be empty?

File: lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs (hunk 4 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/RuntimeName/EvaluatorRequestingDriver.cs]: not fixed yet.

**Comment by afchung** [comment]
@shulmanb still a few minor issues. Thanks!

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs (hunk 16 above)
Comment (shulmanb) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs]: No, but I don't think it should be enforced on the descriptor level.

**Comment by shulmanb** [comment]
Fixed them. Please take a look.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs (hunk 16 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorDescriptorImpl.cs]: Then I don't think ``runtimeName`` should be an optional parameter.

**Comment by afchung** [comment]
Looks good, I'll test and merge.
